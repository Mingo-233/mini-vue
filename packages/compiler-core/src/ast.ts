import { CREATE_ELEMENT_VNODE } from "./runtimeHelpers";
export enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION,
}

export function createVNodeCal(context, tag, props, children) {
  context.addHelper(CREATE_ELEMENT_VNODE);
  return {
    type: NodeTypes.ELEMENT,
    tag,
    props,
    children,
  };
}
