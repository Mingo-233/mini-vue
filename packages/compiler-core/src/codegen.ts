import {
  TO_DISPLAY_STRING,
  CREATE_ELEMENT_VNODE,
  helperMapName,
} from "./runtimeHelpers";
import { NodeTypes } from "./ast";
import { isString } from "../../shared";

export function generate(ast) {
  // 创建上下文
  const context = createCodeGenContext();
  const { push } = context;
  genFunctionPreamble(ast, context);

  // 定义函数名
  const functionName = "render";
  // 定义参数签名
  const args = ["_ctx", "_cache"];
  const signature = args.join(",");
  push(`function ${functionName}(${signature}) {`);

  push(`return `);
  // 拼接代码内容
  genNode(ast.codeGenNode, context);
  push(`}`);

  return {
    code: context.code,
  };
}
function genFunctionPreamble(ast, context) {
  const { push } = context;
  const VueBinging = "Vue";
  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;
  if (ast.helpers.length) {
    push(`const {${ast.helpers.map(aliasHelper).join(",")}} = ${VueBinging}`);
  }
  push(`\n`);

  push(`return `);
}

function createCodeGenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };

  return context;
}
function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;

    default:
      break;
  }
}
function genCompoundExpression(node, context) {
  const { push } = context;
  const child = node.children;
  for (let i = 0; i < child.length; i++) {
    const node = child[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }
  }
}
function genElement(node, context) {
  //example: <div>123</div> -> _createElementVnode('div','null',123)
  const { push, helper } = context;
  const { tag, props, children } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  genNodeList(genNullable([tag, props, children]), context);
  push(`)`);
}
function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }
    // 参数位之间分割，最后一位的时候不用加，所以-1
    if (i < nodes.length - 1) {
      push(",");
    }
  }
}
function genNullable(arr) {
  return arr.map((item) => item || "null");
}
function genText(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  // 去解析插值的content内容，即{{}}里面的内容
  genNode(node.content, context);
  push(`)`);
}

function genExpression(node, context) {
  const { push } = context;
  push(`${node.content}`);
}
