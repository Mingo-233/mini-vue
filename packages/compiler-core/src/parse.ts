// baseParse 函数是解析模板的入口函数，接收模板内容字符串作为参数。

import { NodeTypes } from "./ast";
const enum TagType {
  Start,
  End,
}
export function baseParse(content: string) {
  const context = createParserContext(content);
  const children = parseChildren(context, []);
  return createRoot(children);
}

function parseChildren(context, ancestors) {
  const nodes: any[] = [];
  while (!isEnd(context, ancestors)) {
    const source = context.source;
    let node;
    if (source.startsWith("{{")) {
      node = parseInterpolation(context);
      nodes.push(node);
    } else if (source[0] === "<") {
      if (/[a-z]/i.test(source[1])) {
        node = parseElement(context, ancestors);
        nodes.push(node);
      }
    }
    if (!node) {
      node = parseText(context);
      nodes.push(node);
    }
  }

  return nodes;
}
function parseElement(context, ancestors) {
  const element: any = parseTag(context, TagType.Start);
  ancestors.push(element);
  element.children = parseChildren(context, ancestors);
  ancestors.pop();
  console.log("element", element);
  console.log("context.source", context.source);

  if (isStartsWithEndTag(context.source, element.tag)) {
    // 消费结束标签
    parseTag(context, TagType.End);
  } else {
    throw Error(`缺少结束标签:${element.tag}`);
  }

  return element;
}
function parseText(context) {
  // example: 'hi,{{message}}</div>' 解析到这种情况的时候，应该取的是'hi,'
  let endIndex = context.source.length;
  const endTokens = ["{{", "<"];
  for (let i = 0; i < endTokens.length; i++) {
    const endI = context.source.indexOf(endTokens[i]);
    if (endI !== -1 && endIndex > endI) {
      endIndex = endI;
    }
  }
  const content = parseTextData(context, endIndex);
  return {
    type: NodeTypes.TEXT,
    content: content,
  };
}
function parseTextData(context, length) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}
function parseTag(context, type: TagType) {
  // <div></div>
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  //   ['<div', 'div', index: 0, input: '<div></div>', groups: undefined]
  const tag = match[1]; // 提取标签名。
  // 前进解析上下文，跳过匹配的部分。
  advanceBy(context, match[0].length);
  // 跳过 '>' 字符。
  advanceBy(context, 1);
  if (type === TagType.End) return;
  return {
    type: NodeTypes.ELEMENT,
    tag: tag,
    children: [],
  };
}
function parseInterpolation(context) {
  const openDelimiter = "{{"; // 插值表达式的起始定界符。
  const closeDelimiter = "}}"; // 插值表达式的结束定界符。
  // 查找插值表达式结束定界符的位置。
  const closeIndex = context.source.indexOf(closeDelimiter);
  // 跳过起始定界符。
  advanceBy(context, openDelimiter.length);
  // 计算原始内容的长度。
  const rawContentLength = closeIndex - openDelimiter.length;

  // 提取原始内容。
  const rawContent = parseTextData(context, rawContentLength);
  const content = rawContent.trim();
  advanceBy(context, closeDelimiter.length);
  // 返回解析后的插值节点。
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content,
    },
  };
}
function isEnd(context, ancestors) {
  const s = context.source;
  // 1.判断是否是结束标签
  if (s.startsWith("</")) {
    // Note: 这里为什么用数组去循环找，就是为了找到一个结束标签，保证一定能退出外层的while循环
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (isStartsWithEndTag(s, tag)) {
        return true;
      }
    }
  }
  // 2.source是否存存在
  return !s;
}
// 判断当前这个结束标签和前序的开始标签是否一致
function isStartsWithEndTag(source, tag) {
  return source.startsWith(`</`) && source.slice(2, 2 + tag.length) === tag;
}
// advanceBy 函数用于前进解析上下文的指针。
function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}
// createParserContext 函数创建解析上下文。
function createParserContext(content: string) {
  return {
    source: content,
  };
}

// createRoot 函数创建根节点。
function createRoot(children) {
  return {
    children,
    type: NodeTypes.ROOT,
  };
}
