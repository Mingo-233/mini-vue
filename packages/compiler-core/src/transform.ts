import { TO_DISPLAY_STRING } from "./runtimeHelpers";
import { NodeTypes } from "./ast";
export function transform(ast, options = {}) {
  // 1 深度优先搜索找到目标节点
  // 2 通过外界传入行为，处理修改节点数据
  const context = createTransformContext(ast, options);
  traverseNode(ast, context);
  // 生成codeGenNode属性，供generate code 的时候入口
  createRootCodeGen(ast);
  ast.helpers = [...context.helpers.keys()];
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    addHelper: function (key) {
      context.helpers.set(key, 1);
    },
  };
  return context;
}

function traverseNode(node, context) {
  const nodeTransforms = context.nodeTransforms;
  const exitsFns: any[] = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const fn = nodeTransforms[i];
    const exitsFn = fn(node, context);
    exitsFn && exitsFns.push(exitsFn);
  }
  switch (node.type) {
    case NodeTypes.ELEMENT:
    case NodeTypes.ROOT:
      traverseChildren(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      context.addHelper(TO_DISPLAY_STRING);
      break;

    default:
      break;
  }
  let len = exitsFns.length;
  while (len--) {
    exitsFns[len]();
  }
}
function traverseChildren(node, context) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const child = children[i];
      traverseNode(child, context);
    }
  }
}

function createRootCodeGen(root) {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT) {
    root.codeGenNode = child.codeGenNode;
  } else {
    root.codeGenNode = root.children[0];
  }
}
