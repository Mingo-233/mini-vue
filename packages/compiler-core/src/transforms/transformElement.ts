import { NodeTypes, createVNodeCal } from "../ast";
export function transformElement(node, context) {
  return () => {
    if (node.type === NodeTypes.ELEMENT) {
      const tag = node.tag;
      let props;
      // TODO: ?
      const child = node.children[0];

      node.codeGenNode = createVNodeCal(context, tag, props, child);
    }
  };
}
