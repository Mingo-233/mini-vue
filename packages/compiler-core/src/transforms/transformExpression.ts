import { NodeTypes } from "../ast";

export function transformExpression(node) {
  // 处理插值表达式的content值
  if (node.type === NodeTypes.INTERPOLATION) {
    // 去处理 插值的content内容
    console.log("_ctx_ctx_ctx");

    node.content = processExpression(node.content);
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`;
  return node;
}
