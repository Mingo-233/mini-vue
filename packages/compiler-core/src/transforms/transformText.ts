import { NodeTypes } from "../ast";
import { isText } from "../utils";

export function transformText(node) {
  return () => {
    // 把 普通文字内容和合并，例如文字+插值
    if (node.type === NodeTypes.ELEMENT) {
      // 容器存储
      let currentContainer;
      // 循环children
      let child = node.children;
      for (let i = 0; i < child.length; i++) {
        const node = child[0];
        // 当前节点是否是文字

        if (isText(node)) {
          for (let j = i + 1; j < child.length; j++) {
            const nextNode = child[j];
            // 下一个节点是否是文字
            if (isText(nextNode)) {
              if (!currentContainer) {
                // 初始化容器 - 类型为复合节点

                currentContainer = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [node],
                };
                // 和ast节点关联上
                child[i] = currentContainer;
              }
              // 进行拼接
              currentContainer.children.push("+");
              currentContainer.children.push(nextNode);
              // 在原数组中移除掉刚拼接的一项
              child.splice(j, 1);
              // 因为移除了一项，索引减1，如果不变化，那么后面前移的一项会丢失
              j--;
            } else {
              // 如果不是文字 直接结束循环
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    }
  };
}
