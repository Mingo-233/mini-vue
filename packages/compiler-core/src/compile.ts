import { baseParse } from "./parse";
import { transform } from "./transform";
import { generate } from "./codegen";
import { transformExpression } from "./transforms/transformExpression";
import { transformElement } from "./transforms/transformElement";
import { transformText } from "./transforms/transformText";

export function baseCompile(template: string) {
  const ast = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  return generate(ast);
}
