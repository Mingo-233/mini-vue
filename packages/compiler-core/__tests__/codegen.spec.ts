import { describe, expect, it } from "vitest";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { generate } from "../src/codegen";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformElement } from "../src/transforms/transformElement";
import { transformText } from "../src/transforms/transformText";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("abc");
    transform(ast);
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
  it("element", () => {
    const ast: any = baseParse("<div>hi,{{message}}</div>");
    transform(ast, {
      // nodeTransforms: [transformExpression, transformText, transformElement],
      //Note: 因为后面2个转换函数是退出执行的，所以这边的实际执行顺序是 1transformExpression 2transformText 3transformElement
      // 为什么要做退出执行，是因为 element类型节点在transformText环节结构会改变，内部的差值类型会被并入复合类型中，后面就无法对这个差值类型处理了
      //  所以做尾执行 ，让前面的transformExpression都先走完
      nodeTransforms: [transformExpression, transformElement, transformText],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
