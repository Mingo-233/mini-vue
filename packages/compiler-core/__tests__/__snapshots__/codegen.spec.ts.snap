// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > element 1`] = `
"const {toDisplayString:_toDisplayString,createElementVNode:_createElementVNode} from Vue
return function render(_ctx,_cache) {return _createElementVNode('div','null','hi,'+_toDisplayString(_ctx.message))}"
`;

exports[`codegen > interpolation 1`] = `
"const {toDisplayString:_toDisplayString} from Vue
return function render(_ctx,_cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen > string 1`] = `
"
return function render(_ctx,_cache) {return 'abc'}"
`;
